### ISSUE TITLE
Write all the AMG matrices in a V-cycle to file

### ISSUE THREAD

**Issue by PaulMullowney** [hypre-space/hypre#649]
Hi,
Is there a way to way to write matrices, in a BoomerAMG sequence, to disk. In particular, I want files for the matrices used in the smoothing operators.
-Paul

**Comment by pbauman** [comment]
Hey @PaulMullowney. The (Par)CSRMatrix classes have methods for writing the matrices out. You can extract all the matrices from the `hypre_ParAMGData` (that is extracted from the, e.g., `hypre_PCGData`). You'd get one file for each of the matrices in the CSRMatrix case and nprocs*n_matrices in the ParCSRMatrix case. Is that what you're looking for?

**Comment by PaulMullowney** [comment]
Hi @pbauman  Essentially yes. I was wondering if this had already been implemented somewhere which has already been exposed through some API/option.

**Comment by pbauman** [comment]
Not AFAICT, but the HYPRE developers may know.. I just manually extracted the matrices and dumped them out.

**Comment by liruipeng** [comment]
How about this https://github.com/hypre-space/hypre/blob/07a8def6f8b7dfdffa4208d962f08f5e5d80ccaa/src/parcsr_ls/par_amg_setup.c#L16? but I am not sure if it works with GPUs. If not, I can fix. Thanks!

**Comment by PaulMullowney** [comment]
The issue seems to be that hypre_ParCSRMatrixPrintIJ assume either CPU matrices or GPU matrices in UVM.
Do you agree?

**Comment by liruipeng** [comment]
> The issue seems to be that hypre_ParCSRMatrixPrintIJ assume either CPU matrices or GPU matrices in UVM. Do you agree?

Yes

**Comment by PaulMullowney** [comment]
This worked for me:

````
(base) [mullowne@login2.summit src]$ git diff
diff --git a/src/parcsr_mv/par_csr_matrix.c b/src/parcsr_mv/par_csr_matrix.c
index cf108f904..56940bc95 100644
--- a/src/parcsr_mv/par_csr_matrix.c
+++ b/src/parcsr_mv/par_csr_matrix.c
@@ -672,6 +672,26 @@ hypre_ParCSRMatrixPrintIJ( const hypre_ParCSRMatrix *matrix,
       return hypre_error_flag;
    }

+#if defined(HYPRE_USING_GPU)
+   diag_i =    hypre_TAlloc(HYPRE_Int,     hypre_CSRMatrixNumRows(diag) + 1, HYPRE_MEMORY_HOST);
+   diag_j =    hypre_TAlloc(HYPRE_Int,     hypre_CSRMatrixNumNonzeros(diag), HYPRE_MEMORY_HOST);
+   diag_data = hypre_TAlloc(HYPRE_Complex, hypre_CSRMatrixNumNonzeros(diag), HYPRE_MEMORY_HOST);
+
+   hypre_TMemcpy(diag_i,    hypre_CSRMatrixI(diag),    HYPRE_Int,     hypre_CSRMatrixNumRows(diag) + 1, HYPRE_MEMORY_HOST, hypre_CSRMatrixMemoryLocation(diag));
+   hypre_TMemcpy(diag_j,    hypre_CSRMatrixJ(diag),    HYPRE_Int,     hypre_CSRMatrixNumNonzeros(diag), HYPRE_MEMORY_HOST, hypre_CSRMatrixMemoryLocation(diag));
+   hypre_TMemcpy(diag_data, hypre_CSRMatrixData(diag), HYPRE_Complex, hypre_CSRMatrixNumNonzeros(diag), HYPRE_MEMORY_HOST, hypre_CSRMatrixMemoryLocation(diag));
+
+   num_nonzeros_offd = hypre_CSRMatrixNumNonzeros(offd);
+   if (num_nonzeros_offd)
+   {
+          offd_i =    hypre_TAlloc(HYPRE_Int,     hypre_CSRMatrixNumRows(offd) + 1, HYPRE_MEMORY_HOST);
+          offd_j =    hypre_TAlloc(HYPRE_Int,     hypre_CSRMatrixNumNonzeros(offd), HYPRE_MEMORY_HOST);
+          offd_data = hypre_TAlloc(HYPRE_Complex, hypre_CSRMatrixNumNonzeros(offd), HYPRE_MEMORY_HOST);
+          hypre_TMemcpy(offd_i,    hypre_CSRMatrixI(offd),    HYPRE_Int,     hypre_CSRMatrixNumRows(offd) + 1, HYPRE_MEMORY_HOST, hypre_CSRMatrixMemoryLocation(offd));
+          hypre_TMemcpy(offd_j,    hypre_CSRMatrixJ(offd),    HYPRE_Int,     hypre_CSRMatrixNumNonzeros(offd), HYPRE_MEMORY_HOST, hypre_CSRMatrixMemoryLocation(offd));
+          hypre_TMemcpy(offd_data, hypre_CSRMatrixData(offd), HYPRE_Complex, hypre_CSRMatrixNumNonzeros(offd), HYPRE_MEMORY_HOST, hypre_CSRMatrixMemoryLocation(offd));
+   }
+#else
    diag_data = hypre_CSRMatrixData(diag);
    diag_i    = hypre_CSRMatrixI(diag);
    diag_j    = hypre_CSRMatrixJ(diag);
@@ -683,6 +703,7 @@ hypre_ParCSRMatrixPrintIJ( const hypre_ParCSRMatrix *matrix,
       offd_i    = hypre_CSRMatrixI(offd);
       offd_j    = hypre_CSRMatrixJ(offd);
    }
+#endif

    ilower = row_starts[0] + (HYPRE_BigInt) base_i;
    iupper = row_starts[1] + (HYPRE_BigInt) base_i - 1;
@@ -739,6 +760,19 @@ hypre_ParCSRMatrixPrintIJ( const hypre_ParCSRMatrix *matrix,

    fclose(file);

+#if defined(HYPRE_USING_GPU)
+   hypre_TFree(diag_i,    HYPRE_MEMORY_HOST);
+   hypre_TFree(diag_j,    HYPRE_MEMORY_HOST);
+   hypre_TFree(diag_data, HYPRE_MEMORY_HOST);
+
+   num_nonzeros_offd = hypre_CSRMatrixNumNonzeros(offd);
+   if (num_nonzeros_offd)
+   {
+          hypre_TFree(offd_i,    HYPRE_MEMORY_HOST);
+          hypre_TFree(offd_j,    HYPRE_MEMORY_HOST);
+          hypre_TFree(offd_data, HYPRE_MEMORY_HOST);
+   }
+#endif
    return hypre_error_flag;
 }
````

**Comment by victorapm** [comment]
Hi @PaulMullowney, thanks for you suggested fix. I implemented a similar one in #655 and merged it into master, so now you can call `hypre_ParCSRMatrixPrintIJ` for a matrix living on the device without the need of UVM and it will work. 

Thanks!
